<a href="https://colab.research.google.com/github/NickReeder/Network_Sciences_S23/blob/main/hw3%5Chw3_report.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

All Notebooks must start with code which allows opening the file in Google collab. For example the following code, is responsible for the `Open in Colab` blue button.
```
[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/anwala/teaching-network-science/blob/main/spring-2023/week-1/data_340_02_hw3_report.ipynb)
```
[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/11usZu77_dPlzH2rI2zornhWKJhz0JJJp?usp=sharing)

Ensure that you include the correct file path of your notebook file. Study the location of my [notebook file on Github](https://github.com/anwala/teaching-network-science/blob/main/spring-2023/week-1/data_340_02_hw2_report.ipynb) and above path (`github/anwala/teaching-network-science/blob/main/spring-2023/week-1/data_340_02_hw3_report.ipynb`) I used to produce the blue button. 

# HW3 - Hubs
### Nick Reeder
### Data 340-02, Spring 2023
### Thursday, March 9 2023


# Q1

Draw the OpenFlighs US flight network graph. Ensure the graph is legible and pretty:

- Node labels are visible

- Edge crossings are minimized


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import networkx as nx
import numpy as np
import pandas as pd
%matplotlib inline

In [ ]:


G = nx.read_graphml('openflights_usa.graphml.gz')

nx.write_gexf(G, 'C:\\Users\\Nick\\Dropbox\\PC\\Desktop\\Network_Sciences\\USA_OpenFlights.gexf')


![\label{fig:USA Open Flight Map}](https://drive.google.com/file/d/1giWQ2x5LEf_fWUj7KqBYtmUU9TuEslvk/view?usp=share_link)

[Link to total image](https://drive.google.com/file/d/1giWQ2x5LEf_fWUj7KqBYtmUU9TuEslvk/view?usp=share_link)

[Link to Continental USA image](https://drive.google.com/file/d/1jm0ILkC5lYWsa7yYkjejyGbdHJhH1vpu/view?usp=share_link)

[Link to Alaska image](https://drive.google.com/file/d/1lxztzRbslmoHCxnhtvZGdq2sm7l1_KD1/view?usp=share_link)

[Link to Hawaii image](https://drive.google.com/file/d/1Zea65E29dayMK6Jhwj0qA_BUjAwJExe-/view?usp=share_link) 

## Discussion
I found using Gephi to at first be challenging, but eventually intuitive. I bounced between the USA map and using ForceAtlas, but decided to use the USA Geoplot as it allowed for a more intuitive read of the graph. After setthing the Geoplot I used the NoOverlap to spread some nodes out. Because there were so many nodes and edges, I found minimizing edge crossing to be rather challening. This was another advantage of the USA Geoplot as even if the edge got covered, one could fairly easily see its direction and intuit the start/end points. 

I also manually dragged the Hawaii and Alaska portions closer as it it maintained the idea of the map while making it more readable overall.

Q2 (1 point x 6)
- What is the average number of routes served by each airport in this network?

- What are the top five airports in terms of number of routes?

- How many airports in this network serve only a single route?

- Which airport has the highest closeness centrality?

- Which airport has the highest betweenness centrality?

- Compute the heterogeneity parameter of this network.

In [ ]:

df = pd.DataFrame(G.degree(), columns = ['Airport', 'Degree'])
df['Closeness_centrality'] = nx.closeness_centrality(G).values()
df['Betweenness_centrality'] = nx.betweenness_centrality(G).values()
df.sort_values("Degree", ascending = False, inplace = True)


print("Average number of routes served by each airport is", df["Degree"].mean())

print("Number of airports with only 1 route:", str(len([df.iloc[i,0] for i in range(0,len(df)) if df.iloc[i,1] == 1])))

print("Top 5 airports in temrs of number of routs are:", str([df.iloc[i,0] for i in range(0,5)])[1:-1])

print("Airport with the highest closeness centrality:",  df.sort_values("Closeness_centrality", ascending = False).iloc[0,0])

print("Airport with the highest betweenness centrality:",  df.sort_values("Betweenness_centrality", ascending = False).iloc[0,0])

print("Heterogenity paramater is:", sum((df['Degree'])**2)/len(df) / (df["Degree"].mean())**2)



Average number of routes served by each airport is 10.186813186813186
Number of airports with only 1 route: 156
Top 5 airports in temrs of number of routs are: 'ATL', 'ORD', 'DEN', 'DFW', 'MSP'
Airport with the highest closeness centrality: DEN
Airport with the highest betweenness centrality: ANC
Heterogenity paramater is: 5.347494770144302



## Discussion
In the begining I had a lot of trouble with Network X and struggled to find documentation on it. I eventually figured out the funcions from one of the example notebooks provided, but had already used pandas for all but the heterogenity paramaeter. I am surprised at how low this is given that there are some nodes with exceptionally high degrees such as ATL and many nodes with only degree 1. That said I do not have a good frame of reference for these numbers so 5.3 could be higher than I am giving it credit for.

I also found pandas useful for looking at the network elements wholistically and seeing different features of it. 

# References

*Every report must list the references that you consulted while completing the assignment. If you consulted a webpage, you must include the URL.*

* Pandas documentation, https://pandas.pydata.org/docs/>
